content/news/XXXX-XX-XX-controller-api-declartions.md

```diff
@@ -0,0 +1,51 @@
+title: Controller API declarations for JavaScript mapping developers
+authors: Jörg Wartenberg
+status: draft
+tags: 2.4, controller
+
+
+If the support of a DJ controller requires more than the semantic 1:1 mapping of MIDI codes, Mixxx offers the possibility to use freely programmable Javascript code to implement such more complex functionalities.
+
+The Mixxx internal Application Programming Interface (API) for such controller scripts was previously only partially documented, in the [Mixxx Wiki](https://github.com/mixxxdj/mixxx/wiki/).
+
+This has changed with Mixxx 2.4, now there are formal TypeScript declarations for each individual API class/method. These can be read by any code editor with [Language Server Protocol (LSP)](https://en.wikipedia.org/wiki/Language_Server_Protocol) support, to be visualized to the developer while typing his controller scripting code.
+![Screenshot of documentation from controller API declaration visualized in editor]({static}/images/news/ControllerApiDocuInEditor.png)
+Popular editors and IDEs with LSP support are Atom, Eclipse, Sublime Text, Visual Studio and Visual Studio Code.
+
+In addition to the inline visualization of the documentation, the syntax is also checked against the TypeScript declarations. Usage errors are reported directly when writing the code in the editor.
+![Screenshot of syntax check against controller API declaration visualized in editor]({static}/images/news/ControllerApiSyntaxCheckInEditor.png)
+
+## How to use it?
+Development of mappings and controller scripts is usually done in the subdirectory `controllers` of the [Mixxx Settings Directory](https://manual.mixxx.org/latest/chapters/appendix/settings_directory), while the common files (like `common-controller-scripts.js`) which comes with Mixxx, are store in the subdirectory `controllers` of Mixxx installation directory, or `./mixxx/res/controllers` if you built Mixxx yourself. This is also the location, where the TypeScript files with the declarations of the Mixxx controller API are stored. The LSP language server need to know the location of all of them. To tell the LSP server general settings and the locations of these files, you need to create a file with the name `jsconfig.json` placed next to your JavaScript controller script file, in the `controllers` directory in your `Mixxx Settings Directory`.
+
+The `jsconfig.json` must have a content like this:
+
+    {
+    "compilerOptions": {
+        "target": "es6",
+        "checkJs": true,
+        "lib": [ "ES2016" ]
+    },
+    "include": [
+        "./Traktor-Kontrol-Z2-hid-scripts.js",
```

Comment:
Is this the name of the controller script in development? Is it possible to use wildcards/globs/regex? Otherwise I'd name it `filename_of_your_controller_script.js` or something like that.

Reply:
A wildcard shouldn't be used here, because it would find other mapping scripts in the folder as well. I will clarify this!